G1 support (#69)
drand/tlock: README.md

tlock gives you time based encryption and decryption capabilities by relying on a [drand](https://drand.love/) threshold network.  
It's also a Go library, which is used to implement the `tle` command line tool enabling anybody to leverage timelock encryption.

Our timelock encryption system relies on an "[unchained drand network](https://drand.love/blog/2022/02/21/multi-frequency-support-and-timelock-encryption-capabilities/)".

Working endpoints to access it are, on mainnet:
- https://api.drand.sh/ (US)
- https://api2.drand.sh/ (EU)
- https://api3.drand.sh/ (Asia)
- https://drand.cloudflare.com/ (load-balanced across regions)

On mainnet, only one chainhash supports timelock encryption:
- running with a 3 seconds frequency with signatures on G1: `dbd506d6ef76e5f386f41c651dcb808c5bcbd75471cc4eafa3f4df7ad4e4c493`

> This is a production-ready network with high-availability guarantees. It is considered by the drand team and ran by

Reviewer: what do you mean `considered by the drand team`?

Author: forgot a "fully secure" there :P

tlock gives you time based encryption and decryption capabilities by relying on a [drand](https://drand.love/) threshold network.  
It's also a Go library, which is used to implement the `tle` command line tool enabling anybody to leverage timelock encryption.

Our timelock encryption system relies on an "[unchained drand network](https://drand.love/blog/2022/02/21/multi-frequency-support-and-timelock-encryption-capabilities/)".

Working endpoints to access it are, on mainnet:
- https://api.drand.sh/ (US)
- https://api2.drand.sh/ (EU)
- https://api3.drand.sh/ (Asia)
- https://drand.cloudflare.com/ (load-balanced across regions)

> On mainnet, only one chainhash supports timelock encryption:
> - running with a 3 seconds frequency with signatures on G1: `dbd506d6ef76e5f386f41c651dcb808c5bcbd75471cc4eafa3f4df7ad4e4c493`

Reviewer 2: ```suggestion
On mainnet, the chainhas supporting timelock encryption, with a 3s frequency and signatures on the G1 group is: `dbd506d6ef76e5f386f41c651dcb808c5bcbd75471cc4eafa3f4df7ad4e4c493`
```

> This is a production-ready network with high-availability guarantees. It is considered fully secure by the drand team 

Reviewer: Could mention here that it's been audited. Kudelsk put out a [timelock-encrypted disclosure blog](https://research.kudelskisecurity.com/2023/01/09/announcing-a-timelocked-responsible-disclosure/) ... but it will only be decryptable using the `testnet-unchained-3s` on the old scheme x( we could maybe release our own blog post with the findings and our remediation?

Author: Let's do that later once we have the final version of the report 👍🏻

and ran by the same League of Entropy that has been running drand in production since 2019.

On testnet:
- https://pl-us.testnet.drand.sh/
- https://pl-eu.testnet.drand.sh/
- https://testnet0-api.drand.cloudflare.com/
where we have two networks supporting timelock:
- running with a 3 seconds frequency with signatures on G1: `f3827d772c155f95a9fda8901ddd59591a082df5ac6efe3a479ddb1f5eeb202c`
- running with a 3 seconds frequency with signatures on G2: `7672797f548f3f4748ac4bf3352fc6c6b6468c9ad40ad456a397545c6e2df5bf`
Note these are relying on the League of Entropy **Testnet**, which should not be considered secure.

> You can also spin up a new drand network and run your own, but note that the security guarantees boil down to the trust you can have in your network.

Reviewer: ```suggestion
You can also spin up a new drand network and run your own, but note that the security guarantees boil down to the trust you have in your network.
```

Files can be encrypted using a duration (`--duration/-D`) in which the `encrypted_data` can be decrypted.

Example using the testnet network and a duration of 5 seconds:
```bash
$ tle -n="https://pl-us.testnet.drand.sh/" -c="7672797f548f3f4748ac4bf3352fc6c6b6468c9ad40ad456a397545c6e2df5bf" -D=5s -o=encrypted_data data.txt
```

If a round (`--round/-R`) number is known, it can be used instead of the duration. The data can be decrypted only when that round becomes available in the network.

> Example using the fastnet mainnet network and a given round:

Reviewer: note these parameters can also be left out

```bash
$ tle -n="https://pl-us.testnet.drand.sh/" -c="7672797f548f3f4748ac4bf3352fc6c6b6468c9ad40ad456a397545c6e2df5bf" -D=5s -o=encrypted_data data.txt
```

If a round (`--round/-R`) number is known, it can be used instead of the duration. The data can be decrypted only when that round becomes available in the network.

Example using the fastnet mainnet network and a given round:
```bash
$ tle -n="https://api.drand.sh/" -c="dbd506d6ef76e5f386f41c651dcb808c5bcbd75471cc4eafa3f4df7ad4e4c493" -r=123456 -o=encrypted_data data.txt
```

It is also possible to encrypt the data to a PEM encoded format using the armor (`--armor/-a`) flag:
```bash
$ tle -a -D 20s -o=encrypted_data.PEM data.txt
```

#### Time Lock Decryption

For decryption, it's only necessary to specify the network if you're not using the default one.

Using the default ("fastnet" network on mainnet) and printing on stdout:
```bash
$ tle -d encrypted_data
```

Using the old testnet unchained network and storing the output in a file named "decrypted_data":
```bash
$ tle -d -n="https://pl-us.testnet.drand.sh/" -c="7672797f548f3f4748ac4bf3352fc6c6b6468c9ad40ad456a397545c6e2df5bf"
 -o=decrypted_data encrypted_data
```

> Notice it will overwrite that `decrypted_data` file if it exists already.

Reviewer: ```suggestion
Note it will overwrite the `decrypted_data` file if it already exists.
```
